### ISSUE TITLE
Observability mechanisms

### ISSUE THREAD

**Issue by bgavrilMS** [AzureAD/microsoft-authentication-library-for-go#285]
Please see https://identitydivision.visualstudio.com/DevEx/_git/AuthLibrariesApiReview/pullrequest/4733

The gist of it is: 

- add a few metrics as part of `AuthenticationResult` such as "time spent in cache", "time spent in http" etc.
- small logging improvements
- review telemetry

**Comment by bgavrilMS** [comment]
I agree that this is a critical aspect of MSALs, but it can be added later as it does not imply breaking changes.

**Comment by element-of-surprise** [comment]
I need to read through observability stuff.  In Go you either end up with logs doing observability (bad) or observability that is tied to the Context object, that way it can flow through the entire call chain.  One of the best ways to do it is using OTEL.  Don't know how much OTEL support MSAL's other libraries have.

**Comment by bgavrilMS** [comment]
> I need to read through observability stuff. In Go you either end up with logs doing observability (bad) or observability that is tied to the Context object, that way it can flow through the entire call chain. One of the best ways to do it is using OTEL. Don't know how much OTEL support MSAL's other libraries have.
> 
> One of the keys to GA should be making sure we have Context support on all public methods/interfaces so this is easy to instrument later.

OpenTelemetry support is being considered, we don't have it yet.
I think observability can be added later than this GA.
Absolutely agree about Context support on all public methods.
